packages/react-events/src/FocusScope.js

```diff
@@ -37,95 +35,112 @@ function focusFirstChildEventTarget(
         if (altkey || ctrlKey || metaKey) {
           return;
         }
         const elements = context.getFocusableElementsInScope();
-        const position = elements.indexOf(currentFocusedNode);
+        const position = elements.indexOf(focusedElement);
+        const lastPosition = elements.length - 1;
+        let nextElement = null;
+
         if (shiftKey) {
           if (position === 0) {
             if (props.trap) {
-              focusLastChildEventTarget(context, state);
+              nextElement = elements[lastPosition];
             } else {
+              // Out of bounds
+              context.releaseOwnership();
               return;
             }
           } else {
-            const previousElement = elements[position - 1];
-            previousElement.focus();
-            state.currentFocusedNode = previousElement;
+            nextElement = elements[position - 1];
           }
         } else {
-          if (position === elements.length - 1) {
+          if (position === lastPosition) {
             if (props.trap) {
-              focusFirstChildEventTarget(context, state);
+              nextElement = elements[0];
             } else {
+              // Out of bounds
+              context.releaseOwnership();
               return;
             }
           } else {
-            const nextElement = elements[position + 1];
-            nextElement.focus();
-            state.currentFocusedNode = nextElement;
+            nextElement = elements[position + 1];
+          }
+        }
+        // If this element is possibly inside the scope of another
+        // FocusScope responder or is out of bounds, then we release ownership.
+        if (nextElement !== null) {
+          if (!context.isTargetWithinEventResponderScope(nextElement)) {
+            context.releaseOwnership();
           }
+          nextElement.focus();
+          state.currentFocusedNode = nextElement;
+          ((nativeEvent: any): KeyboardEvent).preventDefault();
         }
-        ((nativeEvent: any): KeyboardEvent).preventDefault();
       }
     }
   },
-  onMount(
+  onRootEvent(
+    event: ReactResponderEvent,
     context: ReactResponderContext,
     props: FocusScopeProps,
     state: FocusScopeState,
   ) {
+    const {target} = event;
+
+    // Handle global trapping
+    if (props.trap) {
+      if (!context.isTargetWithinEventComponent(target)) {
+        const currentFocusedNode = state.currentFocusedNode;
+        if (currentFocusedNode !== null) {
+          currentFocusedNode.focus();
+        } else if (props.autoFocus) {
+          focusFirstChildEventTarget(context, state);
+        }
+      }
+    }
+  },
+  onMount(
+    context: ReactResponderContext,
+    props: FocusScopeProps,
+    state: FocusScopeState,
+  ): void {
     if (props.restoreFocus) {
-      state.nodeToRestore = document.activeElement;
+      state.nodeToRestore = context.getActiveDocument().activeElement;
```

Comment:
I think we'll need to do something like check for `blur` events that have a `relatedTarget` within the event component subtree, that way we can know the last element focused (`blurEvent.target`) before focus moved into the FocusScope instance. This should give the expected result in cases where the FocusScope is mounted well before the focus moves into it.

Reply:
Sounds good. We should do this later at some point.